File: src/leetcode/python/medium/CloneGraph_133.py
class UndirectedGraphNode:
    def __init__(self, x):
        self.label = x
        self.neighbors = []


class CloneGraph_133(object):
    # @param node, a undirected graph node
    # @return a undirected graph node
    def cloneGraph(self, node):
        def clone(n):
            if not n:
                return None
            if n in cache:
                return cache.get(n)
            cp = UndirectedGraphNode(n.label)
            cache[n] = cp
            for neighbor in n.neighbors:
                cp.neighbors.append(clone(neighbor))
            return cp

        cache = {}
        return clone(node)

File: src/leetcode/java/CloneGraph_133.java
package leetcode.java.medium;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class CloneGraph_133 {
  private HashMap<UndirectedGraphNode, UndirectedGraphNode> cache = new HashMap<>();

  public UndirectedGraphNode cloneGraph(UndirectedGraphNode node) {
    if (node == null) return null;
    if (cache.containsKey(node)) return cache.get(node);

    UndirectedGraphNode newNode = new UndirectedGraphNode(node.label);
    cache.put(node, newNode);
    for (UndirectedGraphNode neighbor : node.neighbors) {
      newNode.neighbors.add(cloneGraph(neighbor));
    }
    return newNode;
  }

  class UndirectedGraphNode {
    int label;
    List<UndirectedGraphNode> neighbors;

    UndirectedGraphNode(int x) {
      label = x;
      neighbors = new ArrayList<UndirectedGraphNode>();
    }
  }
}
